draft-ietf-tls-external-psk-importer.md

```diff
@@ -79,156 +80,151 @@ when, and only when, they appear in all capitals, as shown here.
-as defined in {{terminology}}, along with an optional label, and transforms
-it into a set of PSKs and imported identities for use in a connection based on supported HashAlgorithms.
-In particular, for each supported HashAlgorithm 'hash', the importer constructs an ImportedIdentity
-structure as follows:
+A key importer takes as input an EPSK with external identity `external_identity` and base key `epsk`,
+as defined in {{terminology}}, along with an optional context, and transforms it into a set of PSKs
+and imported identities for use in a connection based on supported (target) protocols and KDFs. In particular,
+for each supported target protocol `target_protocol` and KDF `target_kdf`, the importer constructs
+an ImportedIdentity structure as follows:
 
 ~~~
-   struct {
-       opaque external_identity<1...2^16-1>;
-       opaque label<0..2^8-1>;
-       opaque context<0..2^16-1>;
-       HashAlgorithm hash;
-   } ImportedIdentity;
+uint8 KDFID[2];
+
+struct {
+   opaque external_identity<1...2^16-1>;
+   opaque context<0..2^16-1>;
+   uint16 target_protocol;
+   KDFID target_kdf;
+} ImportedIdentity;
 ~~~
 
-[[TODO: An alternative design might combine label and hash into the same field so that future
-protocols which don't have a notion of HashAlgorithm don't need this field.]]
+The list of KDFID values is maintained by IANA as described in {{IANA}}. External PSKs MUST NOT
+be imported for versions of (D)TLS 1.2 or prior versions. See {{rollout}} for discussion on
+how imported PSKs for TLS 1.3 and non-imported PSKs for earlier versions co-exist for incremental
+deployment.
 
-ImportedIdentity.label MUST be bound to the protocol for which the key is imported. Thus,
-TLS 1.3 and QUICv1 {{!I-D.ietf-quic-transport}} MUST use "tls13" as the label. Similarly, TLS 1.2 and
-all prior TLS versions should use "tls12" as ImportedIdentity.label, as well as SHA256 as ImportedIdentity.hash.
-Note that this means future versions of TLS will increase the number of PSKs derived from an external PSK.
+ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.
+If the EPSK is a key derived from some other protocol or sequence of protocols,
+ImportedIdentity.context MUST include a channel binding for the deriving protocols
+{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included
+in ImportedIdentity.context {{CCB}}.
 
-ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
-from some other protocol or sequence of protocols, ImportedIdentity.context MUST include a channel binding for the deriving protocols
-{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included in ImportedIdentity.context [CCB].
+ImportedIdentity.target_protocol MUST be the (D)TLS protocol version for which the
+PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!I-D.ietf-quic-transport}}
```

Comment:
Is this the on-wire version number or the abstract version number?

We identify DTLS 1.3 the same as TLS 1.3, with 0x0304, but we identity DTLS 1.2 in two ways: either 0x0303 or 0xfdfd.  I'm sure people will want to know which is used here.

```suggestion
PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!QUIC=I-D.ietf-quic-transport}}
```

Reply:
This is the on-wire version number. (I left out DTLS since DTLS 1.3 uses 0x0304 and I removed import support for (D)TLS 1.2.)